### ISSUE TITLE
Integrate Mobile App with Password Autofill

### ISSUE THREAD

**Issue by mandusm** [padloc/padloc#250]
Both major mobile Operating Systems now support an API to integrate third party authentication providers for logins within the confines of the OS, like apps and browsers. 

[Android](https://developer.apple.com/documentation/security/password_autofill)
[iOS](https://developer.apple.com/documentation/security/password_autofill)

I haven't looked at the Padloc mobile app code yet, but this is very likely a cordova plugin requirement. A quick google of that did not show any results. So this feature request might be blocked until that is available.

**Comment by MaKleSoft** [comment]
> I haven't looked at the Padloc mobile app code yet, but this is very likely a cordova plugin requirement.

Yep, good assessment. If you're an iOS or Android developer, this is probably the best way to contribute to Padloc right now: build a Cordova plugin that handles autofill integrations. If anyones up for it, let me know so we can discuss requirements and security specs.

**Comment by MaKleSoft** [comment]
@mandusm Both your links point to the iOS documenation btw.

**Comment by Raph33** [comment]
@maklesoft I believe that you're not building browser extensions for Padloc because then Padloc would be vulnerable to attacks on the browsers. Is the same not true for mobile autofill?

**Comment by mandusm** [comment]
> @mandusm Both your links point to the iOS documenation btw.

Oops. Thanks, updated.

**Comment by MaKleSoft** [comment]
@Raph33 It's not quite the same. Browsers are a generally a very insecure environment and developing extensions that handle sensitive data securely is quite difficult. These APIs on the other hand were designed specifically for the use case of password managers and are a likely much easier to implement in a secure way. We'll have to take a closer look to tell for sure but it's definitely worth looking into.

Incidentally, browser extensions (or auto fill on desktop in generall) are not off the table either, we just want to make sure we're doing it right (see https://github.com/padloc/padloc/issues/82).

**Comment by marcoclg19** [comment]
What is the probability of seeing autofill on iOS in the near future?

**Comment by MaKleSoft** [comment]
Not very high, honestly. Mobile is currently powered by Cordova, and implementing auto-fill on iOS and Android would require writing a plugin for each. The thing is that we're planning to migrate mobile to [Tauri](https://tauri.app) as soon as it's ready for mobile and I'd rather not spend more time on the Cordova runtime if we're going to replace it soon. Unfortunately, there is [no definitive timeline](https://github.com/tauri-apps/tauri/issues/843) for Tauri mobile yet, so I can't really say when we'll be able to address this.

**Comment by nmk-freelance** [comment]
I would like to help with the autofill on iOS if it is not being migrated in the near future. Let me know whether I should invest in it.

**Comment by MaKleSoft** [comment]
@nmk-freelance Thanks for the offer! I would off on it a little longer - Tauri still doesn't officially support iOS and Android yet but working with Cordova has become more and more cumbersome in the last months (we still haven't figured out how to build the Android app for the API level 31 target) and it's becoming increasingly clear that we'll have to replace it with **something** soon. Unfortunately with Tauri mobile still in development and no other good alternatives currently available, we'll just have to wait and see...

**Comment by nmk-freelance** [comment]
I have explored the iOS developer documentation. Then I found out that it requires AutoFill Credential Provider extension implementing ASCredentialProviderViewController. It just needs to access the credentials from the web storage.

The autofill view is just a UIViewController subclass with a UIView and can display a web view. But I think it would be easier to make a minimalist native UI displaying just the list. The extension may still be useful when Cordova is replaced. I tried [PassCamp](https://www.passcamp.com/) on iOS and it seemed that it had a separate UI for the autofill view.

It might not be so hard for me to do that. I would like to work on it as it will help users access their credentials faster. @MaKleSoft What would you say?

**Comment by MaKleSoft** [comment]
@nmk-freelance If you're super keen to work on this, go for it! I would recommend you make this a standalone plugin though. Once you have something that works, we can try integrating it. Let me know if you need input on the technical design and security considerations. Keep in mind that if we're going to use it it'll have to hold up to an independent security audit so make sure to keep the security aspect in mind 😉

**Comment by nmk-freelance** [comment]
Definitely, I will pay attention to the security aspect. We will see if it is feasible to put most of the stuff into a plugin. I will let you know if I need to know something about the technical design and security considerations.

**Comment by nmk-freelance** [comment]
@MaKleSoft I found a way to sync data to the iOS extension using the Cordova plugin [NativeStorage](https://github.com/TheCocoaProject/cordova-plugin-nativestorage). I have tested in a new Cordova app.

I could not install the Cordova plugin in Padloc Cordova app. I ran `cordova plugin add cordova-plugin-nativestorage` in the directory padloc/packages/cordova and the error was
```
Failed to fetch plugin cordova-plugin-nativestorage via registry.
Probably this is either a connection problem, or plugin spec is incorrect.
Check your connection and plugin name/version/URL.
CordovaError: Error: npm: Command failed with exit code 1 Error output:
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/@padloc%2fapp - Not found
npm ERR! 404 
npm ERR! 404  '@padloc/app@4.2.0' is not in this registry.
npm ERR! 404 
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.
```
How can I install the plugin in the Cordova app?

**Comment by MaKleSoft** [comment]
@nmk-freelance Yeah, that's a bit of an annoying caveat of the monorepo setup since all packages in the `@padloc` namespace are symlinked directly and most of them are not published. You can work around this by removing the `@padloc/app` dependency temporarily. After you've installed the plugin, make sure to add the dependency back and run `npm ci` from the root of the project. Let me know if that helps!

**Comment by nmk-freelance** [comment]
@MaKleSoft I did as you suggested and it worked. Thanks.

**Comment by nmk-freelance** [comment]
@MaKleSoft I have made a working prototype. The changes are in [this branch](https://github.com/nmk-freelance/padloc/commits/autofill-ios). The vault items are synced to the native side after they are saved to the persistent storage and are deleted when logged out. 

Currently, the iOS extension has to be manually added to the Xcode project after running `npm run cordova:build:ios`. So, I made a separate private repository for the native source files. I have invited you to the repository.

Generally, the steps to add the extension are:
- In the Padloc target, add AutoFill Credential Provider in capabilities.
- Add new target called **Autofill**: File > New > Target… > AutoFill Credential Provider. This should create a new group named **Autofill** in the project.
- In both targets, add App Groups capability.
  - Add a container with the id **group.app.padloc.autofill-ios** to both targets.
- Under the Project navigator, replace all the files in the **Autofill** group with the files in the repository.
  - Add `*.swift` and `MainInterface.storyboard` files to **Autofill** target using the File inspector.

Now it can be built using `npm run cordova:build:ios` or in the Xcode IDE. If you need help with adding the extension, let me know.

**Comment by MaKleSoft** [comment]
@nmk-freelance Ok, so I've followed all the steps and successfully built the app with the extension. Padloc now shows up as an autofill provider!

![IMG_1737](https://user-images.githubusercontent.com/1270940/210213440-c070a1b8-dcbc-4a02-9271-e992b6aff2a2.PNG)

To test it, I created a test entry for Github with the url `https://github.com`. I then went to sign in using Safari and clicked the key icon, which bring up this (unfortunately blank) screen:

![IMG_1738](https://user-images.githubusercontent.com/1270940/210213635-1e6d574e-5df7-40ec-8e66-a08f2e23bc84.PNG)

Any ideas what I might be missing?

**Comment by nmk-freelance** [comment]
@MaKleSoft Please check the App Groups ids. They have to be exactly `group.app.padloc.autofill-ios`.
<img width="1072" alt="Screenshot 2023-01-02 at 16 06 22" src="https://user-images.githubusercontent.com/121597772/210214692-135fcedc-3ebd-4b3b-901e-6434a0ec437d.png">
<img width="1072" alt="Screenshot 2023-01-02 at 16 08 47" src="https://user-images.githubusercontent.com/121597772/210214760-d75679c7-1f67-48f3-a2ab-9ae3bbe9b262.png">

**Comment by MaKleSoft** [comment]
@nmk-freelance Yep, I've made sure of that. Any other ideas?

**Comment by nmk-freelance** [comment]
Okay. Please check the console using Safari inspector. Do you see any logs similar to these? It should appear when you unlock the vault or change the items.

<img width="916" alt="Screenshot 2023-01-02 at 16 12 11" src="https://user-images.githubusercontent.com/121597772/210215451-3ec49cab-e737-44d8-bd70-7a9fb5fa3423.png">

**Comment by nmk-freelance** [comment]
@MaKleSoft Sorry. I am also having the issue. Let me fix.

**Comment by nmk-freelance** [comment]
@MaKleSoft I solved it by running `npm run cordova:build:ios` again.

**Comment by MaKleSoft** [comment]
@nmk-freelance Never mind, I'm an idiot. I clones your repo but never checked out the `autofill-ios` branch 🤦‍♂️ It works now!

**Comment by MaKleSoft** [comment]
@nmk-freelance Ok, this is pretty neat. I'm impressed you put this together so quickly! However, there is one major architectural security issue which won't be trivial to solve... Can you see it?

**Comment by nmk-freelance** [comment]
@MaKleSoft No, I do not see it. What is it?

**Comment by nmk-freelance** [comment]
@MaKleSoft Do you mean the manual work required to put the things together?

**Comment by MaKleSoft** [comment]
@nmk-freelance You're essentially dumping the entire vault data into native storage *unencrypted*. Sensitive data should *never* be stored in plain text "at rest". To implement this securely, you would have to encrypt the data before sending it to native storage (or use the existing encrypted representation of the vault contents) and then decrypt it "on the other side". This means you'll basically have to reimplement the entire key derivation and decryption process in the extension. But then you would have to prompt for the master password every time you open the autofill view, unless you implement biometric authentication as we do in the main app. All of this get's complicated very quickly. Do you see what I mean?

**Comment by nmk-freelance** [comment]
@MaKleSoft I see. I can try hosting the Cordova app in a web view in the extension. When an item is selected, I will pass the data directly to the extension instead of opening that item. That way, it will not store anything in the native storage. I doubt if iOS would grant enough resources to the app in an extension.

**Comment by MaKleSoft** [comment]
@nmk-freelance Yes, that sounds like it could work! We can/should probably create a dedicated view for that, but for now you can simply use the list view with the `host` filter, similar as to how the browser extension does it: https://github.com/padloc/padloc/blob/main/packages/extension/src/app.ts#L46-L63

There is probably still the problem of sharing data storage since I assume any WebView embedded in the autofill extension probably won't share the same web storage. We could solve this problem by replacing web storage with the native storage plugin as a storage backend though (which might have other advantages as well)! Basically what you'll want to do is to create an implementation of the [`Storage` Interface](https://github.com/padloc/padloc/blob/main/packages/extension/src/app.ts#L46-L63) using the native storage plugin and then set that as the storage backend on `CordovaPlatform`. Again, the extension is a good example of that: https://github.com/padloc/padloc/blob/main/packages/extension/src/platform.ts#L6

**Comment by nmk-freelance** [comment]
> There is probably still the problem of sharing data storage since I assume any WebView embedded in the autofill extension probably won't share the same web storage.

@MaKleSoft Yes, that's true.

> We could solve this problem by replacing web storage with the native storage plugin as a storage backend though (which might have other advantages as well)! Basically what you'll want to do is to create an implementation of the [`Storage` Interface](https://github.com/padloc/padloc/blob/main/packages/extension/src/app.ts#L46-L63) using the native storage plugin and then set that as the storage backend on `CordovaPlatform`.

I have never thought of that. It seems more straightforward. The data saved using the Storage interface is already encrypted, right?



> We can/should probably create a dedicated view for that, but for now you can simply use the list view with the `host` filter, similar as to how the browser extension does it: https://github.com/padloc/padloc/blob/main/packages/extension/src/app.ts#L46-L63

Yes. A minimal view sufficient to choose from the items would be nice. Temporarily, I will test the same Cordova app in the extension. 

I am using the Chrome extension in Vivaldi and it seems the same as the full app to me. However, I can right-click and select the passwords from the contextual menu.

**Comment by MaKleSoft** [comment]
> I have never thought of that. It seems more straightforward. The data saved using the Storage interface is already encrypted, right?

Yes, all data stored through the standard storage mechanism is encrypted at rest.

**Comment by nmk-freelance** [comment]
I have tried embedding the iOS Cordova app in the extension but, without any luck 😢. I removed all Cordova extensions except [`cordova-plugin-ionic-webview`](https://github.com/ionic-team/cordova-plugin-ionic-webview). It turned out to be that the plugin heavily references `UIApplication` which is not available to the iOS extension. I also used [Capacitor](https://capacitorjs.com/) and it is the same too. They do not seem very friendly for extensions.

I am thinking of writing some native code implementing ([`WKURLSchemeHandler`](https://developer.apple.com/documentation/webkit/wkurlschemehandler)) to serve the built files of the [Padloc Web Client](https://github.com/padloc/padloc/tree/main/packages/pwa) to the web view in the extension. In place of the plugins, I will create a small bridge to implement the `Storage` interface. Basically, it will be using the Padloc Web Client with some extras.

**Comment by nmk-freelance** [comment]
@MaKleSoft I have successfully implemented `Storage` interface for both the Cordova app and for the extension and bridged the extension to the native side. I tested it in a simulator and it seemed to work. Please try it out from the branch [pwa-ios-autofill](https://github.com/nmk-freelance/padloc/tree/pwa-ios-autofill). I put the steps to build in [README.md](https://github.com/nmk-freelance/padloc/blob/pwa-ios-autofill/packages/cordova/README.md).

I would like to know if iOS grants enough resources to the extension on a real device. Could you please test it on a device too? My free developer account does not support Autofill Provider Extension capability.

**Comment by MaKleSoft** [comment]
It works! It's actually really cool that we can run the full web app in the autofill context, because that even allows us to create passwords on the fly and then autofill them! Again, I'm super impressed you got this to work so quickly!

https://user-images.githubusercontent.com/1270940/211008772-7fcf6d04-2fa2-4b26-9851-49307d4a46fa.MP4

Now we have to figure out how to get biometric unlock to work, since entering the master password every time kind of defeats the whole convenience aspect. Also, the actual app now doesn't load any more, apparently due to a problem with the new storage backend. I'm getting the following error in the web console: `Unhandled Promise Rejection: Write has failed`.

**Comment by nmk-freelance** [comment]
> Again, I'm super impressed you got this to work so quickly!

@MaKleSoft Thanks!

> Now we have to figure out how to get biometric unlock to work, since entering the master password every time kind of defeats the whole convenience aspect.

Yes. I will do this. Can you also see that the labels in the active sessions are different for the extension and the app? Should I do anything about that?

> Also, the actual app now doesn't load any more, apparently due to a problem with the new storage backend. I'm getting the following error in the web console: `Unhandled Promise Rejection: Write has failed`.

Does that happen on the first time or after that? It seems to have successfully saved the data as the extension loaded the items. Do you see any logs that may be helpful in Xcode console?

**Comment by nmk-freelance** [comment]
> Also, the actual app now doesn't load any more, apparently due to a problem with the new storage backend. I'm getting the following error in the web console: `Unhandled Promise Rejection: Write has failed`.

I have never encountered such error. Maybe because I copied and used
```
const cordovaReady = new Promise((resolve) => document.addEventListener("deviceready", resolve));
```
from `packages/cordova/src/platform.ts` in `packages/cordova/src/storage.ts`. That might be the problem, as the names clash. I am not sure if it is valid. Could you please try with a different name?

**Comment by MaKleSoft** [comment]
> Can you also see that the labels in the active sessions are different for the extension and the app? Should I do anything about that?

That's definitely something we should fix eventually but let's not worry about it for now.

> Does that happen on the first time or after that? It seems to have successfully saved the data as the extension loaded the items. Do you see any logs that may be helpful in Xcode console?

Actually, I logged in and created the item directly from the extension window :) Didn't see anything helpful in the logs but I'll do some more thorough debugging later.

**Comment by nmk-freelance** [comment]
> > Also, the actual app now doesn't load any more, apparently due to a problem with the new storage backend. I'm getting the following error in the web console: `Unhandled Promise Rejection: Write has failed`.
> 
> I have never encountered such error. Maybe because I copied and used
> 
> ```
> const cordovaReady = new Promise((resolve) => document.addEventListener("deviceready", resolve));
> ```
> 
> from `packages/cordova/src/platform.ts` in `packages/cordova/src/storage.ts`. That might be the problem, as the names clash. I am not sure if it is valid. Could you please try with a different name?

@MaKleSoft In case you missed.

**Comment by nmk-freelance** [comment]
> Actually, I logged in and created the item directly from the extension window :)

@MaKleSoft I have not done like that before. I will try it.

**Comment by nmk-freelance** [comment]
@MaKleSoft What stuff do I have to implement in the `Platform` interface for biometric unlock? I guess they might be:

- `supportedAuthTypes`
- `biometricKeyStore`
- `_prepareRegisterAuthenticator`
- `_prepareCompleteAuthRequest`
- `platformAuthType` and
- `supportsPlatformAuthenticator`

And I think the implementation from CordovaPlatform can be used as a reference. Is there anything else I should know?

**Comment by nmk-freelance** [comment]
@MaKleSoft When I tried to set up biometric unlock, the following error occurs. The server is running at http://localhost:3000 by `npm run server:start`. Do I have to configure anything?

![Biometric setup failed](https://user-images.githubusercontent.com/121597772/211277051-4834fd5f-fdc0-42a5-b265-0a34125a328c.png)

**Comment by MaKleSoft** [comment]
> @MaKleSoft What stuff do I have to implement in the Platform interface for biometric unlock? I guess they might be:
>
> supportedAuthTypes
> biometricKeyStore
> _prepareRegisterAuthenticator
> _prepareCompleteAuthRequest
> platformAuthType and
> supportsPlatformAuthenticator
>
> And I think the implementation from CordovaPlatform can be used as a reference. Is there anything else I should know?

You got it exactly right I think. Basically you should be able to copy & paste except the `biometricKeyStore` implementation. For that you'll basically have to replace the JavaScript portion of the `fingerprint-aio` Cordova plugin. Not sure if you can reuse the iOS portion of that plugin.

**Comment by MaKleSoft** [comment]
> @MaKleSoft When I tried to set up biometric unlock, the following error occurs. The server is running at http://localhost:3000/ by npm run server:start. Do I have to configure anything?

Have you made the adjustments to the platform implementation yet? If so, you might just need to add `public_key` to `PL_AUTH_TYPES`:

```sh
PL_AUTH_TYPES=email,totp,public_key
```

**Comment by nmk-freelance** [comment]
@MaKleSoft I have successfully implemented the biometric unlocking in the extension 😃. But, it needs to be set up in the extension without setting it up in the main app. The commit: [Added biometric unlocking with a caveat.](https://github.com/nmk-freelance/padloc/commit/6f3ccad12d8a80b0897fdccf85f94d8858575426)

**Comment by MaKleSoft** [comment]
@nmk-freelance Both the extension and base app should be able to [access the same key chain item as part of an app group](https://developer.apple.com/documentation/security/keychain_services/keychain_items/sharing_access_to_keychain_items_among_a_collection_of_apps), so I think as long as we make sure to use the same storage key then setting up biometric auth in one should make it available in the other as well, no? Looks like the Cordova plugin doesn't actually support using a custom key but uses a [fixed one](https://github.com/NiklasMerz/cordova-plugin-fingerprint-aio/blob/master/src/ios/Fingerprint.swift#L246) instead: `__aio_key`. So just make sure you're using the same key in your extension?

**Comment by nmk-freelance** [comment]
@MaKleSoft Actually, I have already tried using the same key. But it failed anyway. If I set up in the extension, it failed in the main app and vice versa.

**Comment by MaKleSoft** [comment]
How did it fail exactly?

**Comment by nmk-freelance** [comment]
It just failed like this.
![Simulator Screen Shot - iPhone 14 Pro - 2023-01-10 at 17 07 14](https://user-images.githubusercontent.com/121597772/211529418-1ff95381-ada2-40d7-9023-bdf0e19cc556.png)

**Comment by nmk-freelance** [comment]
@MaKleSoft The error stems from failed decryption. Is the key generated dependent on `DeviceInfo`?

<img width="1792" alt="Screenshot 2023-01-10 at 17 27 46" src="https://user-images.githubusercontent.com/121597772/211533701-8b9f4f11-158b-47f9-8b0f-036fe02ceef6.png">

**Comment by nmk-freelance** [comment]
@MaKleSoft My bad. I did not share the keychain between the main app and the extension.

**Comment by nmk-freelance** [comment]
@MaKleSoft I have fixed the issue in this commit: [Added Keychain Sharing.](https://github.com/nmk-freelance/padloc/commit/0f004b3b649d13f589690c2005b2577dde16b107)

**Comment by MaKleSoft** [comment]
@nmk-freelance Ok, great! Does this mean biometric auth works across both the main app and extension?

**Comment by nmk-freelance** [comment]
@MaKleSoft Yes, it does.

**Comment by MaKleSoft** [comment]
@nmk-freelance Nice. So it looks like we have a proof of concept! Now the next step would be to automate this entire setup process and integrate the whole thing into our build process. Do you think you can figure out how to do that?

**Comment by nmk-freelance** [comment]
@MaKleSoft I think I can. I guess making some [Cordova hooks](https://cordova.apache.org/docs/en/latest/guide/appdev/hooks/) will do that.

**Comment by nmk-freelance** [comment]
@MaKleSoft There is [cordova-node-xcode](https://github.com/apache/cordova-node-xcode) to modify the Xcode project file. Writing a plugin that modifies the project file will work. But, it seems a bit cumbersome to do so.

The project files built using Cordova are static except the `www` folder and the `dist` folder in the extension, if I get it right. I think it would be easier to commit the whole Xcode workspace and use symlinks for `www` and `dist`. What do you think?

**Comment by MaKleSoft** [comment]
> The project files built using Cordova are static except the www folder and the dist folder in the extension, if I get it right. I think it would be easier to commit the whole Xcode workspace and use symlinks for www and dist. What do you think?

I'd like to avoid adding the entire workspace to version control for various reasons, but mostly because it would changing large parts of our build process. If there is a way to programmatically build the project folder from scratch including the additional setup, that would be preferable. @BrunoBernardino Do you have any thoughts on this?

**Comment by BrunoBernardino** [comment]
@MaKleSoft I agree with you, though I also know it might be very difficult to get CI working well without that versioned. It feels worthy to try, still.

**Comment by nmk-freelance** [comment]
@BrunoBernardino Since Cordova produces the Xcode project as part of the build process, it might be easier to extend the process. @MaKleSoft So, I think [cordova-node-xcode](https://github.com/apache/cordova-node-xcode) is worth trying out.

**Comment by nmk-freelance** [comment]
@MaKleSoft Have you already tried this by any chance? Thank you.
